examples/crypto/SP800-108_counter_KDF/main.c

```diff
@@ -0,0 +1,341 @@
+	if (status != PSA_SUCCESS) {
+		printf("\npsa_mac_sign_setup failed! (Error: %d)\n", status);
+		return status;
+	}
+
+	status =
+		mac_update_fixed_input(&operation, label, label_length, context, context_length, L);
+	if (status != PSA_SUCCESS) {
+		printf("\nmac_update_fixed_input failed! (Error: %d)\n", status);
+		return status;
+	}
+
+	status = psa_mac_sign_finish(&operation, K_0, sizeof(K_0), &output_size);
+	if (status != PSA_SUCCESS) {
+		printf("\npsa_mac_sign_finish failed! (Error: %d)\n", status);
+		return status;
+	}
+
+	/* K_i = CMAC(K_in, [i]_4 || label || 0x00 || context || [L]_4 || K_0) */
+	for (uint32_t i = 1; bytes_produced < output_length; i++) {
+		uint8_t K_i[AES_BLOCK_SIZE] = {};
+		/* Setup a new operation */
+		status = psa_mac_sign_setup(&operation, key_id, PSA_ALG_CMAC);
+		if (status != PSA_SUCCESS) {
+			printf("\npsa_mac_sign_setup - 2 failed! (Error: %d)\n", status);
+			return status;
+		}
+
+		/* counter is needed as 4 byte big-endian */
+		uint32_t counter_be = le32_to_be32(i);
+
+		/* [i]_4  */
+		status = psa_mac_update(&operation, (uint8_t *)&counter_be, sizeof(counter_be));
+		if (status != PSA_SUCCESS) {
+			printf("\npsa_mac_update L failed! (Error: %d)\n", status);
+			return status;
+		}
+
+		/* label || 0x00 || context || [L]_4 */
+		status = mac_update_fixed_input(&operation, label, label_length, context,
+						context_length, L);
+		if (status != PSA_SUCCESS) {
+			printf("\nmac_update_fixed_input failed! (Error: %d)\n", status);
+			return status;
+		}
+
+		/* K_0 */
+		status = psa_mac_update(&operation, K_0, sizeof(K_0));
+		if (status != PSA_SUCCESS) {
+			printf("\npsa_mac_update L failed! (Error: %d)\n", status);
+			return status;
+		}
+
+		status = psa_mac_sign_finish(&operation, K_i, sizeof(K_i), &output_size);
+		if (status != PSA_SUCCESS) {
+			printf("\npsa_mac_sign_finish failed! (Error: %d)\n", status);
+			return status;
+		}
+
+		memcpy(output + bytes_produced, K_i,
+		       min(output_size, output_length - bytes_produced));
+		bytes_produced += output_size;
+	}
+
+	printf("Done\n");
+
+	return PSA_SUCCESS;
+}
+
+void main()
+{
+	psa_status_t status = psa_crypto_init();
+	if (status != PSA_SUCCESS) {
+		printf("psa_crypto_init failed! (Error: %d)\n", status);
+		return;
+	}
+
+	/* Do a SP800-108 counter HMAC kdf with SHA256 */
+	status = sp800_108_counter_hmac_kdf(PSA_ALG_SHA_256, m_input_hmac_key,
+					    sizeof(m_input_hmac_key), m_label, sizeof(m_label),
+					    m_context, sizeof(m_context), OUTPUT_SIZE, m_output,
+					    sizeof(m_output));
+
+	if (status == PSA_SUCCESS) {
+		printf("HMAC derived key:\n");
+		print_hex(m_output, sizeof(m_output));
+	}
+
+	/* Do a SP800-108 counter CMAC kdf */
+	status = sp800_108_counter_cmac_kdf(m_input_cmac_key, sizeof(m_input_cmac_key), m_label,
+					    sizeof(m_label), m_context, sizeof(m_context),
+					    OUTPUT_SIZE, m_output, sizeof(m_output));
+
+	if (status == PSA_SUCCESS) {
+		printf("CMAC derived key:\n");
```

Comment:
We will have to print all the parameters of the algorithm in both to the cases. We will need to show the key and all the other parameters and then present the generated materials.
The point being that we want for someone to run the program and get all the data they need without having to look at the source code.

Reply:
Ohhh I wasn't aware we want that, especially as it adds more boilerplate code I would have seen this as keep it as slim as possible to the KDF protocol, was already unsure about all the error printouts, not really like them